src/ProductManifold.jl

```diff
@@ -0,0 +1,416 @@
+similar(x::ProductArray{ShapeSpec}, ::Type{T}) where {ShapeSpec<:ShapeSpecification, T} = ProductArray(ShapeSpec, similar(x.data, T))
+
+"""
+    ProductMPoint(parts)
+
+A more general but slower representation of points on a product manifold.
+"""
+struct ProductMPoint{TM<:Tuple} <: MPoint
+    parts::TM
+end
+
+ProductMPoint(points...) = ProductMPoint{typeof(points)}(points)
+eltype(x::ProductMPoint) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductMPoint) = ProductMPoint(map(similar, x.parts)...)
+similar(x::ProductMPoint, ::Type{T}) where T = ProductMPoint(map(t -> similar(t, T), x.parts)...)
+
+function proj_product(x::ProductMPoint, i::Integer)
+    return x.parts[i]
+end
+
+"""
+    ProductTVector(parts)
+
+A more general but slower representation of tangent vectors on a product
+manifold.
+"""
+struct ProductTVector{TM<:Tuple} <: TVector
+    parts::TM
+end
+
+ProductTVector(vectors...) = ProductTVector{typeof(vectors)}(vectors)
+eltype(x::ProductTVector) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductTVector) = ProductTVector(map(similar, x.parts)...)
+similar(x::ProductTVector, ::Type{T}) where T = ProductTVector(map(t -> similar(t, T), x.parts)...)
+
+(+)(v1::ProductTVector, v2::ProductTVector) = ProductTVector(map(+, v1.parts, v2.parts)...)
+(-)(v1::ProductTVector, v2::ProductTVector) = ProductTVector(map(-, v1.parts, v2.parts)...)
+(-)(v::ProductTVector) = ProductTVector(map(-, v.parts))
+(*)(a::Number, v::ProductTVector) = ProductTVector(map(t -> a*t, v.parts))
+
+function proj_product(x::ProductTVector, i::Integer)
+    return x.parts[i]
+end
+
+"""
+    ProductCoTVector(parts)
+
+A more general but slower representation of cotangent vectors on a product
+manifold.
+"""
+struct ProductCoTVector{TM<:Tuple} <: CoTVector
+    parts::TM
+end
+
+ProductCoTVector(covectors...) = ProductCoTVector{typeof(covectors)}(covectors)
+eltype(x::ProductCoTVector) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductCoTVector) = ProductCoTVector(map(similar, x.parts)...)
+similar(x::ProductCoTVector, ::Type{T}) where T = ProductCoTVector(map(t -> similar(t, T), x.parts)...)
+
+function proj_product(x::ProductCoTVector, i::Integer)
+    return x.parts[i]
+end
+
+function isapprox(M::ProductManifold, x, y; kwargs...)
+    return all(t -> isapprox(t...; kwargs...), ziptuples(M.manifolds, x.parts, y.parts))
+end
+
+function isapprox(M::ProductManifold, x, v, w; kwargs...)
+    return all(t -> isapprox(t...; kwargs...), ziptuples(M.manifolds, x.parts, v.parts, w.parts))
+end
+
+function representation_size(M::ProductManifold, ::Type{T}) where {T}
+    return (mapreduce(m -> sum(representation_size(m, T)), +, M.manifolds),)
+end
+
+manifold_dimension(M::ProductManifold) = sum(map(m -> manifold_dimension(m), M.manifolds))
+
+struct ProductMetric <: Metric end
+
+function det_local_metric(M::MetricManifold{ProductManifold, ProductMetric}, x::ProductArray)
```

Comment:
Can use `mapreduce`

Reply:
Not really, `mapreduce` doesn't have a variant that accepts multiple collections.

Reply:
It seems you're right. Looks like it's coming in Julia 1.2: https://github.com/JuliaLang/julia/pull/31532.

We can still us `mapreduce` with `zip` for some speedups.

```julia
julia> manifolds = [MetricManifold(Euclidean(3), EuclideanMetric()) for i=1:10];

julia> xlist = [randn(3) for i=1:10];

julia> f(manifolds, xlist) = prod(map(det_local_metric, manifolds, xlist));

julia> g(manifolds, xlist) = mapreduce(x->det_local_metric(x...), *, zip(manifolds, xlist));

julia> @benchmark f(manifolds, xlist)
BenchmarkTools.Trial: 
  memory estimate:  240 bytes
  allocs estimate:  5
  --------------
  minimum time:     93.016 ns (0.00% GC)
  median time:      100.420 ns (0.00% GC)
  mean time:        135.009 ns (19.94% GC)
  maximum time:     4.056 μs (95.89% GC)
  --------------
  samples:          10000
  evals/sample:     951

julia> @benchmark g(manifolds, xlist)
BenchmarkTools.Trial: 
  memory estimate:  64 bytes
  allocs estimate:  3
  --------------
  minimum time:     41.319 ns (0.00% GC)
  median time:      43.431 ns (0.00% GC)
  mean time:        56.510 ns (17.19% GC)
  maximum time:     6.070 μs (98.74% GC)
  --------------
  samples:          10000
```

Reply:
Your benchmark is flawed, `manifolds` and `xlist` should be tuples. When that's the case, maps over multiple collections are optimized much better:
```julia
julia> using Manifolds

julia> using BenchmarkTools

julia> manifolds = ntuple( t-> MetricManifold(Euclidean(3), EuclideanMetric()), Val(10));

julia> xlist = ntuple(t -> randn(3), Val(10));

julia> f(manifolds, xlist) = prod(map(det_local_metric, manifolds, xlist));

julia> g(manifolds, xlist) = mapreduce(x->det_local_metric(x...), *, zip(manifolds, xlist));

julia> @benchmark f(manifolds, xlist)
BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     22.136 ns (0.00% GC)
  median time:      22.786 ns (0.00% GC)
  mean time:        35.699 ns (28.56% GC)
  maximum time:     80.109 μs (99.97% GC)
  --------------
  samples:          10000
  evals/sample:     996

julia> @benchmark g(manifolds, xlist)
BenchmarkTools.Trial: 
  memory estimate:  48 bytes
  allocs estimate:  3
  --------------
  minimum time:     34.425 ns (0.00% GC)
  median time:      35.899 ns (0.00% GC)
  mean time:        50.878 ns (23.76% GC)
  maximum time:     86.106 μs (99.94% GC)
  --------------
  samples:          10000
  evals/sample:     993

```

Reply:
Wow, that's quite the speed-up. Cool!

Reply:
For the case of tuples, Julia can even totally constant-fold `f` (but not `g`):

```julia
julia> @code_native f(manifolds, xlist)
	.text
; ┌ @ REPL[13]:1 within `f'
	movabsq	$139763219890848, %rax  # imm = 0x7F1D291302A0
	vmovsd	(%rax), %xmm0           # xmm0 = mem[0],zero
	retq
	nop
; └
```